ledger/src/blockstore/blockstore_purge.rs

```diff
@@ -144,90 +145,181 @@ impl Blockstore {
+            error!(
+                "Error: {:?} while submitting write batch for slot {:?}",
+                e, slot
+            )
+        })?;
+        Ok(columns_purged)
+    }
+
     /// A helper function to `purge_slots` that executes the ledger clean up.
     /// The cleanup applies to \[`from_slot`, `to_slot`\].
     ///
     /// When `from_slot` is 0, any sst-file with a key-range completely older
     /// than `to_slot` will also be deleted.
+    ///
+    /// Note: slots > `to_slot` that chained to a purged slot are not properly
+    /// cleaned up. This function is not intended to be used if such slots need
+    /// to be replayed.
     pub(crate) fn run_purge_with_stats(
         &self,
         from_slot: Slot,
         to_slot: Slot,
         purge_type: PurgeType,
         purge_stats: &mut PurgeStats,
     ) -> Result<bool> {
-        let mut write_batch = self
-            .db
-            .batch()
-            .expect("Database Error: Failed to get write batch");
+        let mut write_batch = self.db.batch()?;
+
         let mut delete_range_timer = Measure::start("delete_range");
+        let columns_purged = self.purge_range(&mut write_batch, from_slot, to_slot, purge_type)?;
+        delete_range_timer.stop();
+
+        let mut write_timer = Measure::start("write_batch");
+        self.db.write(write_batch).inspect(|e| {
+            error!(
+                "Error: {:?} while submitting write batch for purge from_slot {} to_slot {}",
+                e, from_slot, to_slot
+            )
+        })?;
+        write_timer.stop();
+
+        let mut purge_files_in_range_timer = Measure::start("delete_file_in_range");
+        // purge_files_in_range delete any files whose slot range is within
+        // [from_slot, to_slot].  When from_slot is 0, it is safe to run
+        // purge_files_in_range because if purge_files_in_range deletes any
+        // sst file that contains any range-deletion tombstone, the deletion
+        // range of that tombstone will be completely covered by the new
+        // range-delete tombstone (0, to_slot) issued above.
+        //
+        // On the other hand, purge_files_in_range is more effective and
+        // efficient than the compaction filter (which runs key-by-key)
+        // because all the sst files that have key range below to_slot
+        // can be deleted immediately.
+        if columns_purged && from_slot == 0 {
+            self.purge_files_in_range(from_slot, to_slot);
+        }
+        purge_files_in_range_timer.stop();
+
+        purge_stats.delete_range += delete_range_timer.as_us();
+        purge_stats.write_batch += write_timer.as_us();
+        purge_stats.delete_files_in_range += purge_files_in_range_timer.as_us();
+
+        Ok(columns_purged)
+    }
+
+    fn purge_range(
+        &self,
+        write_batch: &mut WriteBatch,
+        from_slot: Slot,
+        to_slot: Slot,
+        purge_type: PurgeType,
```

Comment:
nit: I think it's cleaner  encapsulation to pass a `should_purge_special_columns` boolean here, and then have a function `PurgeType::should_should_purge_special_columns() -> bool` method on `PurgeType`. This way `purge_slot_cleanup_chaining()` doesn't have to know about `PurgeType` at all

Reply:
> This way purge_slot_cleanup_chaining() doesn't have to know about PurgeType at all

I see the merit of adding the new function `purge_slot_cleanup_chaining()` as this has the unique functionality from `run_purge_with_stats`. However, I'm not sure I see the benefit of the boolean over the enum; can you elaborate?

The enum is part of the public API, so I think it is reasonable to expect someone to know about it. And the `Exact` value of the enum means "go purge the special columns for these slots right now", which is what the boolean would convey.

Reply:
1. It seems like the `PurgeType` was designed specifically for the `run_purge_with_stats` workflow, so it seemed better to keep it isolated there rather than mixing it with this new clean function
2. `purge_slot_cleanup_chaining` as a utility function seems like it should just decide which columns to clean. Seems like cramming the `PurgeType` in there was too high level.